File: engines/godot/addons/controller_client/ControllerBridge.cs
using Godot;
using ControllerClient;

/// <summary>
/// Godot 4 Node that wraps the engine-agnostic ControllerClient.
///
/// Add this node to your scene tree. It will:
///   1. Connect to the sensor WebSocket on <c>_Ready()</c>
///   2. Dispatch queued callbacks every frame in <c>_Process()</c>
///   3. Clean up on <c>_ExitTree()</c>
///
/// Access motion data via the <see cref="LastMotion"/> property, button states
/// via <see cref="IsActionPressed"/>, or subscribe to the Godot signals.
///
/// Example (from another script):
/// <code>
/// var bridge = GetNode&lt;ControllerBridge&gt;("ControllerBridge");
/// GD.Print($"Move: {bridge.LastMotion.Move}, Turn: {bridge.LastMotion.Turn}");
///
/// if (bridge.IsActionPressed("jump"))
///     Jump();
/// </code>
/// </summary>
[GlobalClass]
public partial class ControllerBridge : Node
{
    // ── Inspector Properties ──

    [ExportGroup("Connection")]
    [Export(PropertyHint.None, "WebSocket URL of the sensor relay server")]
    public string Url { get; set; } = "ws://localhost:8765/sensor";

    [ExportGroup("Steering")]
    [Export(PropertyHint.Range, "0.1,1.0,0.01")]
    public float MaxTilt { get; set; } = 0.6f;

    [Export(PropertyHint.Range, "0.0,0.5,0.01")]
    public float DeadZone { get; set; } = 0.05f;

    [Export(PropertyHint.Range, "0.01,1.0,0.01")]
    public float SteeringSmoothing { get; set; } = 0.12f;

    [Export(PropertyHint.Range, "10,360,1")]
    public float TurnSpeedDeg { get; set; } = 120f;

    [ExportGroup("Movement")]
    [Export(PropertyHint.Range, "0.05,2.0,0.05")]
    public float StepImpulse { get; set; } = 0.4f;

    [Export(PropertyHint.Range, "0.5,5.0,0.1")]
    public float MaxMove { get; set; } = 1.5f;

    [Export(PropertyHint.Range, "0.1,10.0,0.1")]
    public float MoveDamping { get; set; } = 2.0f;

    // ── Signals ──

    /// <summary>Emitted each frame with the latest processed motion intent.</summary>
    [Signal]
    public delegate void MotionUpdatedEventHandler(float move, float turn, long timestamp, int rawSteps, float stepsCadence);

    /// <summary>Emitted when a command string is received from the controller.</summary>
    [Signal]
    public delegate void CommandReceivedEventHandler(string command);

    /// <summary>Emitted when the WebSocket connection state changes.</summary>
    [Signal]
    public delegate void ConnectionStateChangedEventHandler(bool connected);

    /// <summary>Emitted when a screenshot result is received.</summary>
    [Signal]
    public delegate void ScreenshotReceivedEventHandler(string filePath, int width, int height);

    /// <summary>Emitted when a GPX export result is received.</summary>
    [Signal]
    public delegate void GpxExportedEventHandler(string filePath, double distanceKm, string duration, string error);

    // ── Public State ──

    /// <summary>The underlying engine-agnostic controller instance.</summary>
    public Controller Controller { get; private set; }

    /// <summary>Exposes the active MotionSettings instance.</summary>
    public MotionSettings Settings { get; private set; }

    /// <summary>Last received motion intent. Read this in <c>_Process()</c> or <c>_PhysicsProcess()</c>.</summary>
    public MotionIntent LastMotion { get; private set; }

    /// <summary>Whether the controller WebSocket is currently connected.</summary>
    public bool Connected { get; private set; }

    // ── Lifecycle ──

    public override void _Ready()
    {
        Settings = new MotionSettings();
        SyncSettings();

        Controller = new Controller(Settings);

        Controller.OnMotion += (intent) =>
        {
            LastMotion = intent;
            EmitSignal(SignalName.MotionUpdated,
                intent.Move, intent.Turn, intent.Timestamp,
                intent.rawSteps, intent.StepsCadence);
        };

        Controller.OnCommand += (command) =>
        {
            EmitSignal(SignalName.CommandReceived, command);
        };

        Controller.OnConnectionStateChanged += (isConnected) =>
        {
            Connected = isConnected;
            GD.Print($"[ControllerBridge] Connection: {(isConnected ? "OPEN" : "CLOSED")}");
            EmitSignal(SignalName.ConnectionStateChanged, isConnected);
        };

        Controller.OnScreenshot += (result) =>
        {
            EmitSignal(SignalName.ScreenshotReceived, result.FilePath, result.Width, result.Height);
        };

        Controller.OnGpxExported += (result) =>
        {
            EmitSignal(SignalName.GpxExported, result.FilePath, result.DistanceKm, result.Duration, result.Error);
        };

        Controller.Connect(Url);
        GD.Print($"[ControllerBridge] Connecting to {Url}...");
    }

    public override void _Process(double delta)
    {
        SyncSettings();
        Controller?.Dispatch();
    }

    /// <summary>
    /// Synchronizes the Godot Node properties to the underlying MotionSettings instance.
    /// This allows changing parameters dynamically at runtime (e.g. from the Inspector or scripts).
    /// </summary>
    public void SyncSettings()
    {
        if (Settings != null)
        {
            Settings.MaxTilt = MaxTilt;
            Settings.DeadZone = DeadZone;
            Settings.SteeringSmoothing = SteeringSmoothing;
            Settings.TurnSpeedDeg = TurnSpeedDeg;
            Settings.StepImpulse = StepImpulse;
            Settings.MaxMove = MaxMove;
            Settings.MoveDamping = MoveDamping;
        }
    }

    public override void _ExitTree()
    {
        Controller?.Dispose();
        Controller = null;
    }

    // ── Convenience API ──

    /// <summary>Check whether a named action button is currently pressed.</summary>
    public bool IsActionPressed(string action)
    {
        return Controller?.Actions.Get(action) ?? false;
    }

    /// <summary>Request a full-monitor screenshot from the PC server.</summary>
    public void CaptureScreen()
    {
        Controller?.CaptureScreen();
    }

    /// <summary>Request a window-only screenshot from the PC server (active window if true).</summary>
    public void CaptureScreen(bool windowMode)
    {
        Controller?.CaptureScreen(windowMode);
    }

    /// <summary>Request a screenshot of a specific window by its title.</summary>
    public void CaptureScreen(string windowTitle)
    {
        Controller?.CaptureScreen(windowTitle);
    }

    /// <summary>Request a vibration on the controller.</summary>
    /// <param name="durationMs">The duration of the vibration in milliseconds (default is 200).</param>
    public void Vibrate(int durationMs = 200)
    {
        Controller?.Vibrate(durationMs);
    }

    /// <summary>Start GPX recording with simulated route at server default origin.</summary>
    public void StartGpx()
    {
        Controller?.StartGpx();
    }

    /// <summary>Start GPX recording at server default origin. If manualLocation is true, character positions must be updated via UpdateGpxLocation.</summary>
    public void StartGpx(bool manualLocation)
    {
        Controller?.StartGpx(manualLocation);
    }

    /// <summary>Start GPX recording with simulated route at specified origin.</summary>
    public void StartGpx(double lat, double lon)
    {
        Controller?.StartGpx(lat, lon);
    }

    /// <summary>Start GPX recording at specified origin. If manualLocation is true, character positions must be updated via UpdateGpxLocation.</summary>
    public void StartGpx(double lat, double lon, bool manualLocation)
    {
        Controller?.StartGpx(lat, lon, manualLocation);
    }

    /// <summary>Updates the current GPX location in manual route mode. Automatically throttled.</summary>
    public void UpdateGpxLocation(double lat, double lon)
    {
        Controller?.UpdateGpxLocation(lat, lon);
    }

    /// <summary>Export the recorded GPX trail to the PC server.</summary>
    public void ExportGpx()
    {
        Controller?.ExportGpx();
    }
}

File: engines/cocos-creator/controller-client/ControllerBridge.ts
import { _decorator, Component } from 'cc';
import { Controller } from './Controller';
import { MotionIntent, ScreenshotResult, GpxExportResult } from './MotionIntent';
import { MotionSettings } from './MotionSettings';

const { ccclass, property } = _decorator;

/**
 * Cocos Creator Component that wraps the engine-agnostic Controller.
 *
 * Drop this component onto any Node in your scene. It will:
 *   1. Connect to the sensor WebSocket on `onLoad()`
 *   2. Dispatch queued callbacks every frame in `update()`
 *   3. Clean up on `onDestroy()`
 *
 * Access the latest motion/command data via the public properties
 * or by assigning callbacks before `onLoad()`.
 *
 * Example (from another component):
 *   ```ts
 *   const bridge = this.node.getComponent(ControllerBridge)!;
 *   bridge.controller.onMotion = (intent) => {
 *       this.node.setPosition(intent.move, 0, 0);
 *   };
 *   ```
 */
@ccclass('ControllerBridge')
export class ControllerBridge extends Component {

    // ── Inspector Properties ──

    @property({ tooltip: 'WebSocket URL of the sensor relay server' })
    url: string = 'ws://localhost:8765/sensor';

    @property({ tooltip: 'Maximum tilt angle (normalized)' })
    maxTilt: number = 0.6;

    @property({ tooltip: 'Dead-zone threshold' })
    deadZone: number = 0.05;

    @property({ tooltip: 'Steering smoothing factor (0–1)' })
    steeringSmoothing: number = 0.12;

    @property({ tooltip: 'Turn speed in degrees per second' })
    turnSpeedDeg: number = 120;

    @property({ tooltip: 'Velocity impulse per step' })
    stepImpulse: number = 0.4;

    @property({ tooltip: 'Maximum forward velocity' })
    maxMove: number = 1.5;

    @property({ tooltip: 'Forward velocity damping factor' })
    moveDamping: number = 2.0;

    // ── Public State ──

    /** The underlying engine-agnostic controller instance. */
    public controller!: Controller;

    /** The active motion settings instance. */
    public settings!: MotionSettings;

    /** Last received motion intent (read in `update()` of other components). */
    public lastMotion: MotionIntent = { move: 0, turn: 0, timestamp: 0, rawSteps: 0, stepsCadence: 0 };

    /** Whether the controller WebSocket is currently connected. */
    public connected: boolean = false;

    /** Fired when the motion intent is updated. */
    public onMotionUpdated: ((intent: MotionIntent) => void) | null = null;
    /** Fired when a command is received. */
    public onCommandReceived: ((command: string) => void) | null = null;
    /** Fired when the WebSocket connection state changes. */
    public onConnectionStateChanged: ((connected: boolean) => void) | null = null;
    /** Fired when a screenshot is saved. */
    public onScreenshotReceived: ((result: ScreenshotResult) => void) | null = null;
    /** Fired when a GPX trail is exported. */
    public onGpxExported: ((result: GpxExportResult) => void) | null = null;

    // ── Lifecycle ──

    onLoad(): void {
        this.settings = new MotionSettings();
        this.syncSettings();

        this.controller = new Controller(this.settings);

        this.controller.onMotion = (intent: MotionIntent) => {
            this.lastMotion = intent;
            this.onMotionUpdated?.(intent);
        };

        this.controller.onConnectionStateChanged = (isConnected: boolean) => {
            this.connected = isConnected;
            console.log(`[ControllerBridge] Connection: ${isConnected ? 'OPEN' : 'CLOSED'}`);
            this.onConnectionStateChanged?.(isConnected);
        };

        this.controller.onCommand = (command: string) => {
            this.onCommandReceived?.(command);
        };

        this.controller.onScreenshot = (result: ScreenshotResult) => {
            this.onScreenshotReceived?.(result);
        };

        this.controller.onGpxExported = (result: GpxExportResult) => {
            this.onGpxExported?.(result);
        };

        this.controller.connect(this.url);
    }

    update(_dt: number): void {
        this.syncSettings();
        this.controller?.dispatch();
    }

    /**
     * Synchronizes the Inspector properties to the underlying MotionSettings instance.
     * This allows changing parameters dynamically at runtime (e.g. from other scripts or the Inspector).
     */
    syncSettings(): void {
        if (this.settings) {
            this.settings.maxTilt = this.maxTilt;
            this.settings.deadZone = this.deadZone;
            this.settings.steeringSmoothing = this.steeringSmoothing;
            this.settings.turnSpeedDeg = this.turnSpeedDeg;
            this.settings.stepImpulse = this.stepImpulse;
            this.settings.maxMove = this.maxMove;
            this.settings.moveDamping = this.moveDamping;
        }
    }

    onDestroy(): void {
        this.controller?.dispose();
    }

    // ── Convenience API ──

    /** Check whether a named action button is currently pressed. */
    isActionPressed(action: string): boolean {
        return this.controller?.actions.get(action) ?? false;
    }

    /** Request a full-monitor screenshot from the PC server. */
    captureScreen(): void;
    /** Request a window-only screenshot from the PC server (active window if true). */
    captureScreen(windowMode: boolean): void;
    /** Request a screenshot of a specific window by its title. */
    captureScreen(windowTitle: string): void;
    captureScreen(arg?: boolean | string): void {
        this.controller?.captureScreen(arg as any);
    }

    /** Request a vibration on the controller. */
    vibrate(durationMs: number = 200): void {
        this.controller?.vibrate(durationMs);
    }

    /** Start GPX recording with simulated route at server default origin. */
    startGpx(): void;
    /** Start GPX recording at server default origin. If manualLocation is true, character positions must be updated via UpdateGpxLocation. */
    startGpx(manualLocation: boolean): void;
    /** Start GPX recording with simulated route at specified origin. */
    startGpx(lat: number, lon: number): void;
    /** Start GPX recording at specified origin. If manualLocation is true, character positions must be updated via UpdateGpxLocation. */
    startGpx(lat: number, lon: number, manualLocation: boolean): void;
    startGpx(arg1?: boolean | number, arg2?: number, arg3?: boolean): void {
        this.controller?.startGpx(arg1 as any, arg2 as any, arg3 as any);
    }

    /** Updates the current GPX location in manual route mode. Automatically throttled. */
    updateGpxLocation(lat: number, lon: number): void {
        this.controller?.updateGpxLocation(lat, lon);
    }

    /** Export the recorded GPX trail to the PC server. */
    exportGpx(): void {
        this.controller?.exportGpx();
    }
}
